### PR TITLE
Add full URL path to socialImage across posts

### PR DESCRIPTION
**Pull request by ltagliaferri**
### What should this PR do?
Resolves [DEVED-8](https://sourcegraph.atlassian.net/jira/software/projects/DEVED/boards/9?selectedIssue=DEVED-8) by adding full URL path to `socialImage` in headers. 

Additionally adds a JPG to `/public/headers` as this image was located off site. 

### Why are we making this change?
In order to enable effective social sharing of our content, we'll need to ensure social cards are working across social media platforms. We noticed that there was an issue with social cards on the main about site blog when relative paths were used as opposed to full URL paths (see [Slack discussion](https://sourcegraph.slack.com/archives/C01KPE23H4M/p1626441467156200)). 

### What are the acceptance criteria? 
* Social images should work as expected
* Site should work as expected

### How should this PR be tested?
The links in the deploy preview need to be checked against a open graph image validator, such as https://www.opengraph.xyz/. 

## Pull request process

**Reviewers**:

1. Test functionality using the criteria above.
2. Offer tips for efficiency, feedback on best practices, and possible alternative approaches and things that may not have been considered.
3. For shorter, "quick" PRs, use your best judgement on #​2.
4. Use a collaborative approach and provide resources and/or context where appropriate.
5. Provide screenshots/grabs where appropriate to show findings during review.

**Reviewees**:

1. Prefer incremental and appropriately-scoped changes.
2. Leave a comment on things you want explicit feedback on.
3. Respond clearly to comments and questions.

## Open Questions following this PR!

1. Should we continue to split up the header into an `.svg` and `.png`/`.jpg`? We were using SVGs for fast loads but those are not compatible with social cards. Is having potentially twice as many assets worth any additional conversion / uploading effort for the load time of the site (is there a noticeable load time difference)? See also [Image Conversion & Optimization](https://github.com/sourcegraph/learn#image-conversion-and-optimization) of README.

2. Is now a good time to move images to a GCP bucket vs. hosting them in this GitHub repo? The risk in moving to GCP I think is if we choose an alternate infrastructure solution and may need to port them over down the line. However, staying on the GitHub repo is not ideal.

### CODE REVIEW

**Comment by marekweb** [comment]
The marketing site has had issues with relative URLs, but on the learn platform we have a feature to automatically turn the relative URLs into absolute ones by adding the full site URL:

https://github.com/sourcegraph/learn/blob/7f2308d45606b423beed1c02fdea5ca6054570ca/components/Layout.tsx#L44-L49

There's a benefit to adding the full URL automatically like this, which is that we can test social previews on URLs which are PR deploy previews. On the other hand if we use full hard-coded URLs in `socialImage` then we can't easily test them because they won't point to the deploy preview URL.

If we find any bugs with this, I'd suggest that we fix any such bugs with the automatic URL handling, and for the above reason I'd recommend that we avoid hardcoding the full URLs.

> Should we continue to split up the header into an .svg and .png/.jpg? We were using SVGs for fast loads but those are not compatible with social cards. Is having potentially twice as many assets worth any additional conversion / uploading effort for the load time of the site (is there a noticeable load time difference)? See also Image Conversion & Optimization of README.

I'm a supporter of using SVGs but I agree that it's best to avoid additional effort for little gain, so I don't oppose the idea of keeping just the PNGs.

> Is now a good time to move images to a GCP bucket vs. hosting them in this GitHub repo? The risk in moving to GCP I think is if we choose an alternate infrastructure solution and may need to port them over down the line. However, staying on the GitHub repo is not ideal.

Maybe we should go ahead with that, and I would vote for having an automated script that takes the `public` directory and uploads it all to storage. If we make sure that we have a platform-agnostic URL (our own subdomain basically) then we could switch to any other CDN in the future. I would also suggest that we set the storage domain as a config value and automatically prefix all assets, so that the storage domain isn't hardcoded in individual URLs.

**Comment by ltagliaferri** [comment]
I'm not sure if this is the most elegant way to implement the open graph images, would appreciate modifications if you think changes should be done.

File: components/Article.tsx [hunk 1]
```diff
@@ -67,6 +67,19 @@ const Article: React.FunctionComponent<Props> = props => {
...
+                <meta name="twitter:image" content={props.socialImage} />
+                ) : (
+                    <meta name="twitter:image" content="https://learn.sourcegraph.com/headers/sourcegraph-learn-header.png" />
+                    )}
+            {props.socialImage ? (
+                <meta property="og:image" content={props.socialImage} />
+                ) : (
+                    <meta property="og:image" content="https://learn.sourcegraph.com/headers/sourcegraph-learn-header.png" />
+                    )}
+
```
Comment (marekweb) [inline components/Article.tsx:82]: The meta tags should go into the `<head>` element along with the existing `og:image` and other meta tags in `Layout`

https://github.com/sourcegraph/learn/blob/1a22942c248e089c62c24544167fd3d9db966031/components/Layout.tsx#L67-L70

In that file we fall back to the default learn header if the page doesn't have its own social image so we can reuse `metaImage` for the `twitter:image`.

https://github.com/sourcegraph/learn/blob/1a22942c248e089c62c24544167fd3d9db966031/components/Layout.tsx#L46

Comment (ltagliaferri) [reply components/Article.tsx:82]: My concern with the current implementation is that it is not behaving as expected. The Twitter use case is the most important one.

**Comment by katjuell** [comment]
Few thoughts here:
- If we are going to start using `NEXT_PUBLIC_URL` then we should add a local `.env` file & perhaps an install script to ensure that folks have the variable set locally (the script can create the file with the correct var). 
- We will need to add it to Netlify (we do not currently have any `env` vars set).
- I am tempted to say that we should hold off on a bucket decision until we are ready to make our larger infra decisions, but could go either way if other folks feel it's a priority.
- I'm a fan of fewer assets for sure, so if that's also how y'all feel then let's go in that direction.

**Comment by katjuell** [comment]
Adding a local `.env` with `NEXT_PUBLIC_URL`, removing the extra `meta` tags from `Article` and including the `MetaTags` in `PageLayout` worked for me:
```
interface Props extends LayoutProps, MetaTags {
    leftColumn?: React.ReactNode
    rightColumn?: React.ReactNode
}
```
Happy to push a commit up if it would be helpful.

File: posts/three-ways-to-search-video.md [hunk 2]
```diff
@@ -2,8 +2,9 @@
 title: Three Ways to Search with Sourcegraph (Video)
 tags: [video, search, sourcegraph]
 author: marek-zaluski
-image: https://i3.ytimg.com/vi/XLfE2YuRwvw/maxresdefault.jpg
+image: https://learn.sourcegraph.com/headers/three-ways-to-search-header.jpg
```
Comment (katjuell) [inline posts/three-ways-to-search-video.md:5]: Does this one need to be converted?

**Comment by marekweb** [comment]
> If we are going to start using NEXT_PUBLIC_URL then we should add a local .env file & perhaps an install script to ensure that folks have the variable set locally (the script can create the file with the correct var).

> We will need to add it to Netlify (we do not currently have any env vars set).

Netlify automatically adds a `DEPLOY_URL` environment variable during build, and we use that value. We just translate it to `NEXT_DEPLOY_URL` so that it's available in the component.

https://github.com/sourcegraph/learn/blob/101e94b438317a74e54efa020aea35e765976b34/next.config.js#L10-L13

> including the MetaTags in PageLayout worked for me

I may be misunderstanding but I think that's not necessary because `Layout` props already contain `metaTags` prop, that's how `Article`

Please do correct me if there's an aspect that I'm missing, but it looks to me like the only change that we need to do for Twitter cards to work is to add `twitter:card`, `twitter:image` (and any other Twitter-specific tags if there are more) into the existing section of meta tags here:

https://github.com/sourcegraph/learn/blob/1a22942c248e089c62c24544167fd3d9db966031/components/Layout.tsx#L67-L70

Assuming the `metaImage` value is working as intended already and is prefixing `DEPLOY_URL` (as far as I know this is working correctly) then we can keep relative URLs in the front matter, avoiding hard-coding the absolute URLs.

File: posts/three-ways-to-search-video.md (hunk 2 above)
Comment (ltagliaferri) [reply posts/three-ways-to-search-video.md:5]: yes

**Comment by katjuell** [comment]
Am I missing something about how this works? 
```
DEPLOY_URL: This URL represents the unique URL for an individual deploy. It starts with a unique ID that identifies the deploy; for example, https://5b243e66dd6a547b4fee73ae--petsof.netlify.app
```
It does not sound like that maps to our `learn.sourcegraph.com` url?

**Comment by katjuell** [comment]
I think we do need that change in `PageLayout` because that wasn't being extended as far as I could tell.
Screen Shot 2021-07-29 at 11 14 15 AM

**Comment by ltagliaferri** [comment]
> Am I missing something about how this works?
> 
> ```
> DEPLOY_URL: This URL represents the unique URL for an individual deploy. It starts with a unique ID that identifies the deploy; for example, https://5b243e66dd6a547b4fee73ae--petsof.netlify.app
> ```
> 
> It does not sound like that maps to our `learn.sourcegraph.com` url?

I believe we had to map things with Terraform

**Comment by marekweb** [comment]
> It does not sound like that maps to our learn.sourcegraph.com url?

You're right. It points to a netlify.app URL, which goes to the correct deploy and loads the resource correctly but it doesn't take into account the fact that `learn.sourcegraph.com` is our external URL. Here's what I see in `og:image` in prod right now. 

https://6102ba244dc9010007de4701--sourcegraph-learn.netlify.app/headers/sourcegraph-learn-header.png

I don't think this is a problem for social previews because they work fine with this URL but it would be better if we could have some other environment variable that tells us that we're on `learn.sourcegraph.com` on prod - I'm not sure if Netlify provides that.

**Comment by katjuell** [comment]
I think we should move to using an `.env` structure not embedded in the config: https://nextjs.org/docs/basic-features/environment-variables
We can then add those variables to Netlify
Screen Shot 2021-07-29 at 11 34 31 AM

**Comment by katjuell** [comment]
Could someone point me to the Terraform PR? Plus anything else to do with top-level routing? Just so I have the context.

**Comment by Princesso** [comment]
In my experience, for images to work for twitter OG, you need to add the twitter meta tag @ltagliaferri

File: posts/three-ways-to-search-video.md (hunk 2 above)
Comment (ltagliaferri) [reply posts/three-ways-to-search-video.md:5]: Actually, I think we don't need the image, just the social image? Not sure if we need a banner on this, @marekweb what do you think?

**Comment by ltagliaferri** [comment]
Addressed in #134

**Pull request closed without merging**